sunbeam-python/sunbeam/plugins/ldap/plugin.py

```diff
@@ -0,0 +1,421 @@
+        try:
+            run_sync(
+                self.jhelper.wait_until_active(
+                    self.model,
+                    apps,
+                    timeout=APPLICATION_DEPLOY_TIMEOUT,
+                )
+            )
+        except (JujuWaitException, TimeoutException) as e:
+            LOG.warning(str(e))
+            return Result(ResultType.FAILED, str(e))
+        return Result(ResultType.COMPLETED)
+
+
+class AddLDAPDomainStep(BaseStep, JujuStepHelper):
+    """Generic step to enable OpenStack application using Terraform"""
+
+    def __init__(
+        self,
+        tfhelper: TerraformHelper,
+        jhelper: JujuHelper,
+        plugin: OpenStackControlPlanePlugin,
+        charm_config: str,
+    ) -> None:
+        """Constructor for the generic plan.
+
+        :param tfhelper: Terraform helper pointing to terraform plan
+        :param jhelper: Juju helper with loaded juju credentials
+        :param plugin: Plugin that uses this plan to perform callbacks to
+                       plugin.
+        """
+        super().__init__(
+            f"Enable OpenStack {plugin.name}",
+            f"Enabling OpenStack {plugin.name} application",
+        )
+        self.tfhelper = tfhelper
+        self.jhelper = jhelper
+        self.plugin = plugin
+        self.model = OPENSTACK_MODEL
+        self.charm_config = charm_config
+        self.client = Client()
+
+    def run(self, status: Optional[Status] = None) -> Result:
+        """Apply terraform configuration to deploy openstack application"""
+        config_key = self.plugin.get_tfvar_config_key()
+
+        try:
+            tfvars = read_config(self.client, config_key)
+        except ConfigItemNotFoundException:
+            tfvars = {}
+        tfvars.update(self.plugin.set_tfvars_on_enable())
+        if tfvars.get("ldap_apps"):
+            tfvars["ldap_apps"][self.charm_config["domain-name"]] = self.charm_config
+        else:
+            tfvars["ldap_apps"] = {self.charm_config["domain-name"]: self.charm_config}
+        self.tfhelper.write_tfvars(tfvars)
+        update_config(self.client, config_key, tfvars)
+
+        try:
+            self.tfhelper.apply()
+        except TerraformException as e:
+            return Result(ResultType.FAILED, str(e))
+        charm_name = "keystone-ldap-{}".format(self.charm_config["domain-name"])
+        apps = ["keystone", charm_name]
+        LOG.debug(f"Application monitored for readiness: {apps}")
+        try:
+            run_sync(
+                self.jhelper.wait_until_active(
+                    self.model,
+                    apps,
+                    timeout=APPLICATION_DEPLOY_TIMEOUT,
+                )
+            )
+        except (JujuWaitException, TimeoutException) as e:
+            LOG.warning(str(e))
+            return Result(ResultType.FAILED, str(e))
+
+        return Result(ResultType.COMPLETED)
+
+
+class LDAPPlugin(OpenStackControlPlanePlugin):
```

Comment:
I dont see a need to make the baseclass OpenStackControlPlanePlugin. I believe BasePlugin should be good enough.
The command is not of type enable/disable and follows repo plugin format.

Reply:
Ignore the above comment, i need to look at the PR closely

Reply:
Done